### PR TITLE
pkg/trace: Merge duplicate entries after normalization

### PR DESCRIPTION
**Pull request by piochelepiotr**
After normalization, we can have some duplicate entries in the stats.

### CODE REVIEW

File: pkg/trace/agent/agent.go [hunk 1]
```diff
@@ -317,6 +321,25 @@ func (a *Agent) convertStats(in pb.ClientStatsPayload, lang, tracerVersion strin
...
 
+func mergeDuplicates(s pb.ClientStatsBucket) {
+	indexes := make(map[stats.Aggregation]int, len(s.Stats))
+	for i, g := range s.Stats {
+		a := stats.NewAggregationFromGroup("", "", "", g)
+		if j, ok := indexes[a]; ok {
+			s.Stats[j].Hits += g.Hits
+			s.Stats[j].Errors += g.Errors
+			s.Stats[j].Duration += g.Duration
+			s.Stats[j].TopLevelHits += g.TopLevelHits
```
Comment (furmmon) [inline pkg/trace/agent/agent.go:332]: we only reset Hits, errors, duration
TopLevelHits are separate

File: pkg/trace/agent/agent.go [hunk 2]
```diff
@@ -296,17 +296,27 @@ func (a *Agent) convertStats(in pb.ClientStatsPayload, lang, tracerVersion strin
 	in.Lang = lang
 	for i, group := range in.Stats {
 		n := 0
+		updated := false
 		for _, b := range group.Stats {
-			normalizeStatsGroup(&b, lang)
+			if normalizeStatsGroup(&b, lang) {
```
Comment (piochelepiotr) [inline pkg/trace/agent/agent.go:301]: not sure if the best way to check if a change was made is to update all the modifier functions.
Maybe we can just run `merge` for each payload.

File: pkg/trace/filters/replacer.go [hunk 3]
```diff
@@ -50,7 +50,9 @@ func (f Replacer) Replace(trace pb.Trace) {
 }
 
 // ReplaceStatsGroup applies the replacer rules to the given stats bucket group.
-func (f Replacer) ReplaceStatsGroup(b *pb.ClientGroupedStats) {
+func (f Replacer) ReplaceStatsGroup(b *pb.ClientGroupedStats) bool {
```
Comment (gbbr) [inline pkg/trace/filters/replacer.go:53]: Can you please update the documentation for these methods to specify what the return `bool` is?
```
... and reports whether the input was altered.
```
Or something like that.

File: pkg/trace/agent/agent.go [hunk 4]
```diff
@@ -296,17 +296,27 @@ func (a *Agent) convertStats(in pb.ClientStatsPayload, lang, tracerVersion strin
 	in.Lang = lang
 	for i, group := range in.Stats {
 		n := 0
+		updated := false
```
Comment (gbbr) [inline pkg/trace/agent/agent.go:299]: nit: but we usually write
```suggestion
		var changed bool
```
Seems like `changed` may be a better word?

**Review by gbbr — COMMENTED** [review]
This is a really hack workaround which we should not be too proud off 😛

**Comment by piochelepiotr** [comment]
> This is a really hack workaround which we should not be too proud off 😛

Agreed! The faster it's gone the better.

File: pkg/trace/agent/agent.go (hunk 2 above)
Comment (gbbr) [reply pkg/trace/agent/agent.go:301]: Yes I think that should be fine, it's not like this happens so many times as to warrant an actual performance improvement. Let's start with "nice" code and avoid premature optimisations. Thanks for thinking of that.

**Review by raphaelgavache — APPROVED** [review]

**Comment by gbbr** [comment]
@piochelepiotr this LGTM, but I think your suggestion also ran through my mind: let's drop the returned `bool` and just run merge every time. Should be fine on this low volume, WDYT?

**Comment by piochelepiotr** [comment]
> @piochelepiotr this LGTM, but I think your suggestion also ran through my mind: let's drop the returned `bool` and just run merge every time. Should be fine on this low volume, WDYT?

@gbbr sounds good to me, update. It makes the code easier. @furmmon what do you think?

**Comment by piochelepiotr** [comment]
using this PR instead: https://github.com/DataDog/datadog-agent/pull/7918 because a rebase screwed the reviewers. Sorry for the people that were pinged.

**Pull request closed without merging**